Sources/_StringProcessing/Algorithms/Matching/Matches.swift

```diff
@@ -183,13 +183,106 @@ extension BidirectionalCollection {
 
 // MARK: Regex algorithms
 
+@available(SwiftStdlib 5.7, *)
+struct RegexMatchesCollection<Output> {
+  let base: Substring
+  let regex: Regex<Output>
+  let startIndex: Index
+  
+  init(base: Substring, regex: Regex<Output>) {
+    self.base = base
+    self.regex = regex
+    self.startIndex = base.firstMatch(of: regex).map(Index.match) ?? .end
+  }
+}
+
+@available(SwiftStdlib 5.7, *)
+extension RegexMatchesCollection: Collection {
+  enum Index: Comparable {
+    case match(Regex<Output>.Match)
```

Comment:
Storing the non-trivial element in the index is a design smell, albeit with the stdlib's reliance on `Collection` instead of iterators or true sequences. I don't know of a better workaround though.

Reply:
Agreed

Reply:
I added a custom iterator so we won't have to pay this cost in most contexts.

Reply:
The iterator doesn't hold on to more than the first match, so it could/should be more efficient than `IndexingIterator`.